### PR TITLE
[tests] Changed JSON string assertions to data structure assertions #109

### LINKED ISSUE #109: [tests] Change JSON string assertions to data structure assertions
We should write tests against the NetJSON data structure as a lua table, not against its JSON string representation which is fragile and could change in the future if the underlying libraries we depend on for JSON encoding and decoding change.

Eg:

```diff
diff --git a/openwisp-monitoring/tests/test_interfaces.lua b/openwisp-monitoring/tests/test_interfaces.lua
index 96e4048..5cf02a6 100644
--- a/openwisp-monitoring/tests/test_interfaces.lua
+++ b/openwisp-monitoring/tests/test_interfaces.lua
@@ -1,6 +1,7 @@
 package.path = package.path .. ";../files/lib/openwisp-monitoring/?.lua"
 
 local luaunit = require('luaunit')
+local cjson = require('cjson')
 
 local address_data = require('test_files/address_data')
 local interface_data = require('test_files/interface_data')
@@ -176,9 +177,9 @@ end
 
 function TestNetJSON.test_virtual_interface_type()
   local netjson_file = assert(loadfile('../files/sbin/netjson-monitoring.lua'))
-  local netjson = netjson_file('*')
-  luaunit.assertNotNil(string.find(netjson,
-    '"type":"virtual","mtu":65536,"txqueuelen":1000,"name":"wg0"'))
+  local netjson = cjson.decode(netjson_file('*'))
+  luaunit.assertEquals(netjson["interfaces"][4]["type"], "virtual")
+  luaunit.assertEquals(netjson["interfaces"][4]["name"], "wg0")
 end
 
 os.exit(luaunit.LuaUnit.run())
```

### PR DESCRIPTION
**Pull request by Aryamanz29**
Closes #109

### CODE REVIEW

File: openwisp-monitoring/tests/test_neighbors.lua [hunk 1]
```diff
@@ -54,12 +54,12 @@ function TestNetJSON.test_neighbors()
...
+  local netjson_string = require('netjson-monitoring')
+  local netjson = cjson.decode(netjson_string)
   luaunit.assertNotNil(test_file_dir .. 'ip_neigh.txt')
-  luaunit.assertNotNil(string.find(netjson, '"mac":"bc:0f:9a:17:5a:5c"'))
-  luaunit.assertNotNil(string.find(netjson, '"ip":"fe80::bfca:28ed:f368:6cbc"'))
-  luaunit.assertNotNil(string.find(netjson, '"interface":"eth1"'))
-
+  luaunit.assertEquals(netjson['neighbors'][1]["interface"], "eth1")
+  luaunit.assertNotNil(netjson['neighbors'][3]["ip"], "fe80::bfca:28ed:f368:6cbc")
+  luaunit.assertEquals(netjson['neighbors'][3]["mac"], "bc:0f:9a:17:5a:5c")
```
Comment (nemesisdesign) [inline openwisp-monitoring/tests/test_neighbors.lua:62]: same here

File: openwisp-monitoring/tests/test_wifi.lua [hunk 2]
```diff
@@ -104,31 +101,37 @@ function TestWifi.test_invert_rx_tx()
...
+  luaunit.assertEquals(netjson["interfaces"][1]["wireless"]["ssid"], "meshID")
+  luaunit.assertEquals(netjson["interfaces"][1]["wireless"]["frequency"], 5200)
+  luaunit.assertEquals(netjson["interfaces"][1]["wireless"]["frequency"], 5200)
+  luaunit.assertEquals(netjson["interfaces"][3]["wireless"]["tx_power"], 20)
+  luaunit.assertEquals(netjson["interfaces"][3]["wireless"]["mode"], "access_point")
+  luaunit.assertEquals(netjson["interfaces"][4]["wireless"]["tx_power"], 20)
+  luaunit.assertEquals(netjson["interfaces"][4]["wireless"]["signal"], -67)
+  luaunit.assertEquals(netjson["interfaces"][4]["wireless"]["ssid"], "meshID")
+  luaunit.assertEquals(netjson["interfaces"][4]["wireless"]["clients"][1]["vht"], true)
+  luaunit.assertEquals(netjson["interfaces"][5]["wireless"]["tx_power"], 20)
```
Comment (nemesisdesign) [inline openwisp-monitoring/tests/test_wifi.lua:117]: same here

File: openwisp-monitoring/tests/test_wifi.lua [hunk 3]
```diff
@@ -104,31 +101,37 @@ function TestWifi.test_invert_rx_tx()
...
+  luaunit.assertEquals(netjson["interfaces"][1]["statistics"]["tx_errors"], 0)
+  luaunit.assertEquals(netjson["interfaces"][1]["statistics"]["tx_bytes"],
+    151599685066)
+  luaunit.assertEquals(netjson["interfaces"][3]["statistics"]["tx_errors"], 0)
+  luaunit.assertEquals(netjson["interfaces"][3]["statistics"]["rx_packets"], 2367515)
+  luaunit.assertEquals(netjson["interfaces"][5]["statistics"]["tx_errors"], 0)
+  luaunit.assertEquals(netjson["interfaces"][5]["statistics"]["rx_packets"], 198)
+  luaunit.assertEquals(netjson["interfaces"][5]["statistics"]["rx_bytes"], 25967)
+  luaunit.assertEquals(netjson["interfaces"][5]["statistics"]["tx_bytes"], 531641723)
+  luaunit.assertEquals(netjson["interfaces"][5]["statistics"]["tx_packets"], 2367747)
```
Comment (nemesisdesign) [inline openwisp-monitoring/tests/test_wifi.lua:134]: same here

**Review by nemesisdesign — CHANGES_REQUESTED** [review]

File: openwisp-monitoring/tests/test_dhcp.lua [hunk 4]
```diff
@@ -69,11 +69,12 @@ function TestDhcp.test_dhcp_leases()
...
-  luaunit.assertNotNil(string.find(netjson, '"mac":"e8:6a:64:3e:4a:3c"'))
-  luaunit.assertNotNil(string.find(netjson, '"client_id":"01:e8:6a:64:3e:4a:3c"'))
-  luaunit.assertNotNil(string.find(netjson, '"ip":"192.168.1.136"'))
-  luaunit.assertNotNil(string.find(netjson, '"expiry":1620788343'))
+  local netjson_string = require('netjson-monitoring')
+  local netjson = cjson.decode(netjson_string)
+  luaunit.assertEquals(netjson["dhcp_leases"][1]["ip"], "192.168.1.136")
+  luaunit.assertEquals(netjson["dhcp_leases"][1]["expiry"], 1620788343)
+  luaunit.assertEquals(netjson["dhcp_leases"][1]["mac"], "e8:6a:64:3e:4a:3c")
+  luaunit.assertEquals(netjson["dhcp_leases"][1]["client_id"], "01:e8:6a:64:3e:4a:3c")
```
[the remark this answers is not in the document]
Comment (Aryamanz29) [reply openwisp-monitoring/tests/test_dhcp.lua:75]: You're right, we should avoid changing the assertion order of existing `unittests`. I have configured some `lua-formatter` extensions in my code editor, and I believe it restructured the code based on the lua table index. I have reversed this change in [`c97b3f2` (#115)](https://github.com/openwisp/openwrt-openwisp-monitoring/pull/115/commits/c97b3f2f2932e3593efea685c1a256005891ee9a)

**Review by nemesisdesign — APPROVED** [review]

**Pull request merged**